All v3 updates

jeffmay opened the pull request:
- Split project into common source code projects
- Add support for Play 2.6 with Scala 2.12.6 and 2.11.12
- Add support for Play 2.7 with Scala 2.13.1
- Added back formatMap support using safer `readsMap[K: KeyReads, V: Reads]` and `writesMap[K: KeyWrites, V: Writes]`
- Add unit tests for `Format.empty`
- Updated README

@rallyhealth/engineers @vsrally

jeffmay committed:
All v3 updates with support for Play 2.7 and Scala 2.13

- Play 2.7 / Scala 2.13 upgrades
- Moved code into shared common projects
- Add back implicit formatMap with proper key encoding
- Clean up common project a bit
- Upgrade scoverage plugin
- Prepare code to avoid binary breaking changes
- Fix some incompatibilities between Scala 2.13 and Scala 2.11
- Upgrade sbt to latest version
- Add workaround to Travis CI to avoid issues with sbt cross-compilation
  confusion

jeffmay commented on build.sbt:
```diff
@@ -1,16 +1,25 @@
 
-scalaVersion in ThisBuild := "2.11.11"
+ThisBuild / semVerEnforceAfterVersion := Some("3.0.0")
+//gitVersioningSnapshotLowerBound in ThisBuild := "3.0.0"
```
Use one or the other, and remove comment.

jeffmay commented:
retest this please

jeffmay commented on README.md:
```diff
@@ -3,7 +3,7 @@
 # Play Json Ops
 
-Augments the [Play Json library](https://www.playframework.com/documentation/2.5.x/ScalaJson) with some helpful
+Augments the [Play Json library](https://www.playframework.com/documentation/2.6.x/ScalaJson) with some helpful
```
Use 2.7.x

usufruct99 replied:
probably need to update all the other 2.6 references then...?

jeffmay replied:
This library cross-compiles to 2.6 as well as 2.7, but we should use the latest version in our documentation.

jeffmay commented on play-json-tests-common/src/main/scala/play/api/libs/json/scalacheck/package.scala:
```diff
@@ -0,0 +1,20 @@
+  /**
+    * An extension class that is used when parallel collections are not available for this Scala version.
+    *
+    * This is needed because the parallel collections module has not been released for Scala 2.13.0-M4.
```
Is this still true for Scala `2.13.1`?

jeffmay commented on play-json-tests-common/src/test/scala/play/api/libs/json/ops/JsonTransformSpec.scala:
```diff
@@ -33,11 +33,11 @@ with JsonImplicits {
-    forAll { (obj: JsObject) =>
-      val topLevelPaths: Seq[JsPath] = obj.fields.map(__ \ _._1)
+    forAll { obj: JsObject =>
+      val topLevelPaths: Seq[JsPath] = Seq(obj.fields.toArray: _*).map(__ \ _._1) // required for Scala 2.13 support
```
This is no longer needed since we split 2.7 into a separate artifact

vsrally commented:
@jeffmay  I am no longer in the world of `web development` so  cannot review. let me know if you need a stamp

jeffmay committed:
Update versions table in README.md

kapkaev commented on build.sbt:
```diff
@@ -20,84 +26,156 @@ def commonProject(id: String): Project = {
+def scalaVersions(includePlayVersion: String): Seq[String] = includePlayVersion match {
+  case Play_2_5 => Seq(Scala_2_11)
+  case Play_2_6 => Seq(Scala_2_12, Scala_2_11)
+  case Play_2_7 => Seq(Scala_2_13)
```
```suggestion
  case Play_2_7 => Seq(Scala_2_11 , Scala_2_13)
```

jeffmay replied:
~~Unfortunately, this version cannot be cross-compiled. I would have to create a separate Scala 2.11 version.~~

Actually, I think I might be able to make this work.

jeffmay replied:
Fixed.

kapkaev commented on play-json-ops-common/src/main/scala/play/api/libs/json/ops/DurationFormat.scala:
```diff
@@ -67,7 +68,17 @@ trait StringDurationFormat extends ImplicitDurationReads with ImplicitDurationWr
+
+  override implicit val finiteDurationReads: Reads[FiniteDuration] = new Reads[FiniteDuration] {
+    override def reads(json: JsValue): JsResult[FiniteDuration] = json.validate[(Long, String)] flatMap {
+      case (x, unit) =>
```
What is `x`?

jeffmay replied:
The scalar value

kapkaev replied:
Can we just add types like `case (x: X, unit: Unit) =>`.

jeffmay committed:
Support Play 2.7 / Scala 2.11 and Scala 2.12

kapkaev commented on .travis.yml:
```diff
@@ -5,7 +5,9 @@ jdk:
-  - sbt clean coverage +test coverageReport && sbt coverageAggregate
+  - sbt clean coverage test \
+    +play26-json-tests-sc13/test \
+    coverageReport && sbt coverageAggregate
```
```suggestion
    sbt coverageAggregate
```
https://github.com/scoverage/sbt-scoverage#notes-on-upgrading-to-version-160

jeffmay replied:
Tight!

jeffmay committed:
Add updates to readme

jeffmay committed:
Support cross-compiling Scala 2.11 and Scala 2.12 for Play 2.7

jeffmay committed:
Update .travis.yml

- Remove redundant coverageReport
- Use Scala 2.13.1
- Fix code coverage capturing
- Fix token in travis CI
